api/v1/gatewayapi_types.go

```diff
@@ -0,0 +1,129 @@
+// Copyright (c) 2024 Tigera, Inc. All rights reserved.
+/*
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1
+
+import (
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// GatewayAPISpec has fields that can be used to customize our GatewayAPI support.
+type GatewayAPISpec struct {
+	// Allow optional customization of the gateway controller deployment.
+	GatewayControllerDeployment *GatewayControllerDeployment `json:"gatewayControllerDeployment,omitempty"`
+}
+
+//+kubebuilder:object:root=true
+
+type GatewayAPI struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec GatewayAPISpec `json:"spec,omitempty"`
+}
+
+//+kubebuilder:object:root=true
+
+type GatewayAPIList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []GatewayAPI `json:"items"`
+}
+
+func init() {
+	SchemeBuilder.Register(&GatewayAPI{}, &GatewayAPIList{})
+}
+
+// Optional customization of the gateway controller deployment.
```

Comment:
nit: While the documentation in this file is well written, it violates the linter in Goland and https://go.dev/doc/comment for not starting the comments of exported variables/funcs with the name of the exported variables/funcs.

Reply:
I'm not sure about this.  It seems to me like an unnatural restriction.  Can you configure Goland not to expect this?

Reply:
We certainly could but goland is following golang comment standards where every exported function comment should start with the function name. I think this has to do with generating go docs.

Reply:
Can you provide a reference?

Reply:
[This](https://go.dev/blog/godoc) blog post states the convention:

```
Notice this comment is a complete sentence that begins with the name of the element it describes. This important convention allows us to generate documentation in a variety of formats, from plain text to HTML to UNIX man pages, and makes it read better when tools truncate it for brevity, such as when they extract the first line or sentence.
```

Looking at the updated documentation though [here](https://go.dev/doc/comment) that the older blog post references I don't see any mention of that, however, all the examples of the exported functions use this convention.

Also to clarify, this isn't goland exactly, this is the golang linter that's catching these (which goland itself uses).

I think there's lots of posts about this if you google it, maybe there's a reason why it's not needed anymore but I find comments supporting that it helps doc generation.

Personally I think it would be best to just stick with this convention unless there's some compelling reason not to, I had thought that this was already an adopted convention at tigera.